### PR TITLE
Adds per-column, query-time index skip option

### PR DESCRIPTION
**Pull request by egalpin**
Closes https://github.com/apache/pinot/issues/12355.  Needs tests but opening for early feedback.

cc @kishoreg @Jackie-Jiang

### CODE REVIEW

File: pinot-common/src/main/java/org/apache/pinot/common/utils/config/QueryOptionsUtils.java [hunk 1]
```diff
@@ -145,6 +149,33 @@ public static boolean isSkipScanFilterReorder(Map<String, String> queryOptions)
     return "false".equalsIgnoreCase(queryOptions.get(QueryOptionKey.USE_SCAN_REORDER_OPTIMIZATION));
   }
 
+  @Nullable
+  public static Map<String, Set<FieldConfig.IndexType>> getIndexSkipConfig(Map<String, String> queryOptions) {
+    // Example config:  indexSkipConfig='col1=inverted,range&col2=inverted'
```
Comment (egalpin) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/config/QueryOptionsUtils.java:154]: should I look for an OSS URL param parser?  Not complex code particularly, but also not a unique problem to need to parse url params

**Comment by gortiz** [comment]
It LGTM. As said in the linked issue, my main concern is on the language to use. I think URL args is fine, but maybe we would need to have a bigger picture because the language we decide to use here is what we should use in future options with the same type

File: pinot-common/src/main/java/org/apache/pinot/common/utils/config/QueryOptionsUtils.java [hunk 2]
```diff
@@ -145,6 +149,33 @@ public static boolean isSkipScanFilterReorder(Map<String, String> queryOptions)
...
+    if (indexSkipConfigStr == null) {
+      return null;
+    }
+
+    String[] perColumnIndexSkip = indexSkipConfigStr.split("&");
+    Map<String, Set<FieldConfig.IndexType>> indexSkipConfig = new HashMap<>();
+    for (String columnConf : perColumnIndexSkip) {
+      String[] conf = columnConf.split("=");
+      if (conf.length != 2) {
+        continue;
```
Comment (gortiz) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/config/QueryOptionsUtils.java:165]: Shouldn't we fail in this case?

Comment (egalpin) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/config/QueryOptionsUtils.java:165]: I think so, ya.  I feel like failed parsing is better to end up as a failed query than a successful query without the intended options set.

Comment (egalpin) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/config/QueryOptionsUtils.java]: Actually it looks like there isn't precedent for throwing exceptions from within `InstancePlanMakerImplV2#applyQueryOptions`, so the "easiest" would be a `RuntimeException`

**Comment by egalpin** [comment]
@gortiz anything that needs attention at the moment? I want to be sure I've addressed any comments and can't find any unaddressed at the moment.

**Review by gortiz — APPROVED** [review]

File: pinot-common/src/main/java/org/apache/pinot/common/utils/config/QueryOptionsUtils.java [hunk 3]
```diff
@@ -145,6 +149,34 @@ public static boolean isSkipScanFilterReorder(Map<String, String> queryOptions)
...
 
+  @Nullable
+  public static Map<String, Set<FieldConfig.IndexType>> getIndexSkipConfig(Map<String, String> queryOptions) {
+    // Example config:  indexSkipConfig='col1=inverted,range&col2=inverted'
+    String indexSkipConfigStr = queryOptions.get(QueryOptionKey.INDEX_SKIP_CONFIG);
+    if (indexSkipConfigStr == null) {
+      return null;
+    }
+
+    String[] perColumnIndexSkip = indexSkipConfigStr.split("&");
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/config/QueryOptionsUtils.java:160]: Suggest using `StringUtils.split(indexSkipConfigStr, '&')` which is more lightweight. The current API will try to do regex match. Same for other split calls

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java [hunk 4]
```diff
@@ -361,6 +361,7 @@ public static class QueryOptionKey {
         public static final String SERVER_RETURN_FINAL_RESULT = "serverReturnFinalResult";
         // Reorder scan based predicates based on cardinality and number of selected values
         public static final String AND_SCAN_REORDERING = "AndScanReordering";
+        public static final String INDEX_SKIP_CONFIG = "indexSkipConfig";
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:364]: (personal preference) I feel `skipIndexes` is more consistent with other option keys

**Review by Jackie-Jiang — COMMENTED** [review]
Sorry for the late comments. Can you also help add some documentation for this new query option?

**Comment by egalpin** [comment]
@Jackie-Jiang I'll make a follow-up PR to address these by end of week

**Comment by npawar** [comment]
Do we have documentation for this @egalpin ?

**Comment by egalpin** [comment]
@npawar Ah, thank you for the reminder!  No, there's no formal docs for this yet.

**Pull request merged**